internal/component/loki/secretfilter/secretfilter.go

```diff
@@ -0,0 +1,321 @@
+			for _, r := range c.Rules {
+				// To find the secret within the text captured by the regex (and avoid being too greedy), we can use the 'secretGroup' field in the gitleaks.toml file.
+				// But it's rare for regexes to have this field set, so we can use a simple heuristic in other cases.
+				//
+				// There seems to be two kinds of regexes in the gitleaks.toml file
+				// 1. Regexes that only match the secret (with no submatches). E.g. (?:A3T[A-Z0-9]|AKIA|ASIA|ABIA|ACCA)[A-Z0-9]{16}
+				// 2. Regexes that match the secret and some context (or delimiters) and have one submatch (the secret itself). E.g. (?i)\b(AIza[0-9A-Za-z\\-_]{35})(?:['|\"|\n|\r|\s|\x60|;]|$)
+				//
+				// For the first case, we can replace the entire match with the redaction string.
+				// For the second case, we can replace the first submatch with the redaction string (to avoid redacting something else than the secret such as delimiters).
+				for _, occ := range r.regex.FindAllStringSubmatch(entry.Line, -1) {
+					// By default, the secret is the full match group
+					secret := occ[0]
+
+					// If a secretGroup is provided, use that instead
+					if r.secretGroup > 0 && len(occ) > r.secretGroup {
+						secret = occ[r.secretGroup]
+					} else {
+						// If not and there are two submatches, the first one is the secret
+						if len(occ) == 2 {
+							secret = occ[1]
+						}
+					}
+
+					// Check if the secret is in the allowlist
+					var allowRule *AllowRule = nil
+					// First check the rule-specific allowlist
+					for _, a := range r.allowlist {
+						if a.Regex.MatchString(secret) {
+							allowRule = &a
+							break
+						}
+					}
+					// Then check the global allowlist
+					if allowRule == nil {
+						for _, a := range c.AllowList {
+							if a.Regex.MatchString(secret) {
+								allowRule = &a
+								break
+							}
+						}
+					}
+					// If allowed, skip redaction
+					if allowRule != nil {
+						level.Info(c.opts.Logger).Log("msg", "secret in allowlist", "rule", r.name, "source", allowRule.Source)
+						continue
+					}
+
+					var redactWith = "<REDACTED-SECRET:" + r.name + ">"
+					if c.args.RedactWith != "" {
+						redactWith = c.args.RedactWith
+						redactWith = strings.ReplaceAll(redactWith, "$SECRET_NAME", r.name)
+						redactWith = strings.ReplaceAll(redactWith, "$SECRET_HASH", hashSecret(secret))
+					}
+					if c.args.PartialMask > 0 {
+						redactWith = secret[:c.args.PartialMask] + redactWith
+					}
+
+					entry.Line = strings.ReplaceAll(entry.Line, secret, redactWith)
+				}
+			}
+
+			for _, f := range c.fanout {
+				select {
+				case <-ctx.Done():
+					return nil
+				case f.Chan() <- entry:
+				}
+			}
+		}
+	}
+}
+
+func hashSecret(secret string) string {
+	hasher := sha1.New()
+	hasher.Write([]byte(secret))
+	return fmt.Sprintf("%x", hasher.Sum(nil))
+}
+
+// Update implements component.Component.
+func (c *Component) Update(args component.Arguments) error {
```

Comment:
Update should likely include the work your doing in new, ie update should handle them changing the toml file or any other field. So maybe a shared func for recreating the internal representation.